File: desearch_bot/records.py
"""A domain's and a sitemap's crawl state, as the bucket store keeps it."""

from __future__ import annotations

from collections.abc import Iterable
from datetime import datetime, timedelta, timezone
from typing import TYPE_CHECKING

from .buckets import sitemap_id
from .schedule import DEFAULT_INTERVAL, Trust
from .states import State
from .visit import Known, KnownSitemap, SitemapUpdate

if TYPE_CHECKING:
    from .loop import DomainWrite


def new_domain(
    rank: int | None,
    tld_group: str | None,
    categories: Iterable[str] = (),
    state: State = State.NEW,
    reason: str | None = None,
    due: datetime | None = None,
) -> dict:
    """A domain the process has not visited yet."""
    return {
        "state": state.value,
        "reason": reason,
        "failures": 0,
        "due": _epoch(due),
        "rank": rank,
        "group": tld_group,
        "categories": sorted(categories),
        "urls": 0,
    }


def known(host: str, domain: dict, sitemaps: Iterable[tuple[str, dict]]) -> Known:
    """What a visit needs to know about a domain before it starts."""
    return Known(
        host=host,
        state=State(domain["state"]),
        failures=domain.get("failures", 0),
        last_ok_at=moment(domain.get("ok")),
        robots_checked_at=moment(domain.get("robots")),
        robots_allows=domain.get("allows"),
        crawl_delay=domain.get("delay"),
        language=domain.get("lang"),
        categories=frozenset(domain.get("categories") or ()),
        sitemaps={url: _known_sitemap(url, record) for url, record in sitemaps},
        canonical_host=domain.get("canonical"),
    )


def written_domain(previous: dict, write: DomainWrite, url_count: int) -> dict:
    """The domain record after a visit: robots fields only when robots was read."""
    visit = write.visit
    record = dict(previous)
    record.update(
        state=write.state.value,
        reason=write.reason,
        failures=write.failures,
        due=_epoch(write.next_due_at),
        ok=_epoch(write.last_ok_at),
        checked=_epoch(write.checked_at),
        canonical=write.canonical_host,
        urls=url_count,
    )
    if visit.robots_read:
        record.update(
            robots=_epoch(write.checked_at),
            robots_status=visit.robots_status,
            allows=visit.robots_allows,
            delay=visit.crawl_delay,
        )
    for key, value in (
        ("lang", visit.language),
        ("declared", visit.declared_lang),
        ("chars", visit.home_chars),
    ):
        if value is not None:
            record[key] = value
    return record


def written_sitemap(previous: dict | None, update: SitemapUpdate, at: datetime) -> dict:
    """The sitemap record after a read; its counts change only when its content did."""
    record = dict(previous or {})
    record.update(
        id=update.id,
        depth=update.depth,
        parent=update.parent_id,
        status=update.status,
        error=update.error,
        etag=update.etag,
        modified=update.last_modified,
        hash=update.content_hash,
        trust=update.trust.value,
        index_lastmod=update.index_lastmod,
        interval=int(update.interval.total_seconds()),
        next=_epoch(update.next_check_at),
        fetched=_epoch(at),
    )
    if update.kind:
        record["kind"] = update.kind
    if update.changed:
        record.update(
            urls=update.url_count, children=update.child_count, changed=_epoch(at)
        )
    record.setdefault("urls", 0)
    return record


def unread_sitemap(url: str, depth: int, parent: int | None) -> dict:
    """A sitemap found in an index that a later visit will read."""
    return {
        "id": sitemap_id(url),
        "depth": depth,
        "parent": parent,
        "status": "ok",
        "trust": Trust.UNKNOWN.value,
        "interval": int(DEFAULT_INTERVAL.total_seconds()),
        "next": None,
        "urls": 0,
    }


def url_count(sitemaps: Iterable[dict]) -> int:
    """URLs a domain lists across the sitemaps that last answered."""
    return sum(
        record.get("urls", 0) for record in sitemaps if record.get("status") == "ok"
    )


def _known_sitemap(url: str, record: dict) -> KnownSitemap:
    return KnownSitemap(
        record["id"],
        url,
        record.get("kind"),
        record.get("depth", 0),
        record.get("parent"),
        record.get("etag"),
        record.get("modified"),
        record.get("hash"),
        timedelta(seconds=record.get("interval", DEFAULT_INTERVAL.total_seconds())),
        moment(record.get("next")),
        Trust(record.get("trust", Trust.UNKNOWN.value)),
        record.get("index_lastmod"),
        record.get("urls", 0),
    )


def _epoch(value: datetime | None) -> int | None:
    return None if value is None else int(value.timestamp())


def moment(value: int | None) -> datetime | None:
    """A stored Unix timestamp as a UTC time."""
    return None if value is None else datetime.fromtimestamp(value, timezone.utc)

File: desearch-bot-rs/src/records.rs
//! A domain's and a sitemap's crawl state, as the bucket store keeps them.

use indexmap::IndexMap;
use serde_json::{json, Value};

use crate::buckets::{sitemap_id, Json};
use crate::crawl::DomainWrite;
use crate::schedule::{Trust, DEFAULT_INTERVAL, SECOND};
use crate::states::State;
use crate::visit::{Known, KnownSitemap, SitemapUpdate};

/// A stored whole number, tolerating a float written for it.
pub fn int(value: Option<&Value>) -> Option<i64> {
    let value = value?;
    value.as_i64().or_else(|| value.as_f64().map(|f| f as i64))
}

/// A stored Unix timestamp as epoch microseconds.
pub fn moment(value: Option<&Value>) -> Option<i64> {
    int(value).map(|seconds| seconds * SECOND)
}

fn seconds(micros: Option<i64>) -> Value {
    micros.map_or(Value::Null, |us| Value::from(us.div_euclid(SECOND)))
}

fn text(value: Option<&Value>) -> Option<String> {
    value.and_then(Value::as_str).map(str::to_string)
}

/// A domain the process has not visited yet.
pub fn new_domain(
    rank: Option<i64>,
    group: Option<&str>,
    categories: &[String],
    state: State,
    reason: Option<&str>,
    due: Option<i64>,
) -> Json {
    let mut categories = categories.to_vec();
    categories.sort();
    let record = json!({
        "state": state.as_str(),
        "reason": reason,
        "failures": 0,
        "due": seconds(due),
        "rank": rank,
        "group": group,
        "categories": categories,
        "urls": 0,
    });
    record.as_object().cloned().unwrap_or_default()
}

/// What a visit needs to know about a domain before it starts.
pub fn known(host: &str, domain: &Json, sitemaps: &IndexMap<String, Json>) -> Option<Known> {
    Some(Known {
        host: host.to_string(),
        state: State::parse(domain.get("state")?.as_str()?)?,
        failures: int(domain.get("failures")).unwrap_or(0),
        last_ok_at: moment(domain.get("ok")),
        robots_checked_at: moment(domain.get("robots")),
        robots_allows: domain.get("allows").and_then(Value::as_bool),
        crawl_delay: domain.get("delay").and_then(Value::as_f64),
        language: text(domain.get("lang")),
        categories: domain
            .get("categories")
            .and_then(Value::as_array)
            .map(|all| all.iter().filter_map(|c| c.as_str().map(str::to_string)).collect())
            .unwrap_or_default(),
        sitemaps: sitemaps.iter().filter_map(|(url, record)| Some((url.clone(), known_sitemap(url, record)?))).collect(),
        canonical_host: text(domain.get("canonical")),
    })
}

fn known_sitemap(url: &str, record: &Json) -> Option<KnownSitemap> {
    Some(KnownSitemap {
        id: int(record.get("id"))?,
        url: url.to_string(),
        kind: text(record.get("kind")),
        depth: int(record.get("depth")).unwrap_or(0),
        parent_id: int(record.get("parent")),
        etag: text(record.get("etag")),
        last_modified: text(record.get("modified")),
        content_hash: text(record.get("hash")),
        interval: record.get("interval").and_then(Value::as_f64).map_or(DEFAULT_INTERVAL, |s| (s * 1e6).round() as i64),
        next_check_at: moment(record.get("next")),
        trust: record.get("trust").and_then(Value::as_str).and_then(Trust::parse).unwrap_or(Trust::Unknown),
        index_lastmod: text(record.get("index_lastmod")),
        url_count: int(record.get("urls")).unwrap_or(0),
    })
}

/// The domain record after a visit: robots fields only when robots was read.
pub fn written_domain(previous: &Json, write: &DomainWrite, url_count: i64) -> Json {
    let visit = &write.visit;
    let mut record = previous.clone();
    record.insert("state".into(), write.state.as_str().into());
    record.insert("reason".into(), write.reason.clone().into());
    record.insert("failures".into(), write.failures.into());
    record.insert("due".into(), seconds(write.next_due_at));
    record.insert("ok".into(), seconds(write.last_ok_at));
    record.insert("checked".into(), seconds(Some(write.checked_at)));
    record.insert("canonical".into(), write.canonical_host.clone().into());
    record.insert("urls".into(), url_count.into());
    if visit.robots_read {
        record.insert("robots".into(), seconds(Some(write.checked_at)));
        record.insert("robots_status".into(), visit.robots_status.into());
        record.insert("allows".into(), visit.robots_allows.into());
        record.insert("delay".into(), visit.crawl_delay.into());
    }
    if let Some(language) = &visit.language {
        record.insert("lang".into(), language.clone().into());
    }
    if let Some(declared) = &visit.declared_lang {
        record.insert("declared".into(), declared.clone().into());
    }
    if let Some(chars) = visit.home_chars {
        record.insert("chars".into(), chars.into());
    }
    record
}

/// The sitemap record after a read; its counts change only when its content did.
pub fn written_sitemap(previous: Option<&Json>, update: &SitemapUpdate, at: i64) -> Json {
    let mut record = previous.cloned().unwrap_or_default();
    record.insert("id".into(), update.id.into());
    record.insert("depth".into(), update.depth.into());
    record.insert("parent".into(), update.parent_id.into());
    record.insert("status".into(), update.status.into());
    record.insert("error".into(), update.error.clone().into());
    record.insert("etag".into(), update.etag.clone().into());
    record.insert("modified".into(), update.last_modified.clone().into());
    record.insert("hash".into(), update.content_hash.clone().into());
    record.insert("trust".into(), update.trust.as_str().into());
    record.insert("index_lastmod".into(), update.index_lastmod.clone().into());
    record.insert("interval".into(), (update.interval / SECOND).into());
    record.insert("next".into(), seconds(update.next_check_at));
    record.insert("fetched".into(), seconds(Some(at)));
    if let Some(kind) = &update.kind {
        record.insert("kind".into(), kind.clone().into());
    }
    if update.changed {
        record.insert("urls".into(), update.url_count.into());
        record.insert("children".into(), update.child_count.into());
        record.insert("changed".into(), seconds(Some(at)));
    }
    record.entry("urls").or_insert(0.into());
    record
}

/// A sitemap found in an index that a later visit will read.
pub fn unread_sitemap(url: &str, depth: i64, parent: Option<i64>) -> Json {
    let record = json!({
        "id": sitemap_id(url),
        "depth": depth,
        "parent": parent,
        "status": "ok",
        "trust": Trust::Unknown.as_str(),
        "interval": DEFAULT_INTERVAL / SECOND,
        "next": null,
        "urls": 0,
    });
    record.as_object().cloned().unwrap_or_default()
}

/// URLs a domain lists across the sitemaps that last answered.
pub fn url_count<'a>(sitemaps: impl Iterator<Item = &'a Json>) -> i64 {
    sitemaps.filter(|s| s.get("status").and_then(Value::as_str) == Some("ok")).map(|s| int(s.get("urls")).unwrap_or(0)).sum()
}
